src/cryptonote_core/blockchain_based_list.cpp

```diff
@@ -0,0 +1,278 @@
+{
+  size_t src_list_size = src_list.size();
+
+  if (items_count > src_list_size)
+    items_count = src_list_size;
+
+  static const size_t block_hash_items_count = sizeof(block_hash.data) / sizeof(*block_hash.data);
+
+  size_t step                 = size_t(log2(double(src_list.size())) / 8) + 1,
+         max_iterations_count = block_hash_items_count / step;   
+
+  if (items_count > max_iterations_count)
+    items_count = max_iterations_count;
+
+  decltype(&block_hash.data[0]) hash_ptr = &block_hash.data[0];
+
+  for (size_t i=0; i<items_count; i++, hash_ptr += step)
+  {
+    size_t base_index = extract_index(hash_ptr, step);
+
+    for (size_t offset=0;; offset++)
+    {    
+      if (offset == src_list_size)
+        return; //all supernodes have been selected
+
+      size_t               supernode_index     = (offset + base_index) % src_list_size;
+      const supernode& src_desc            = src_list[supernode_index];
+      const std::string&   supernode_public_id = src_desc.supernode_public_id;
+
+      if (selected_supernodes_cache.find(supernode_public_id) != selected_supernodes_cache.end())
+        continue; //supernode has been already selected
+
+      dst_list.push_back(src_desc);
+
+      selected_supernodes_cache.insert(supernode_public_id);
+
+      break;
+    }
+  }
+}
+
+void BlockchainBasedList::apply_block(uint64_t block_height, const crypto::hash& block_hash, StakeTransactionStorage& stake_txs_storage)
+{
+  if (block_height <= m_block_height)
+    return;
+
+  if (block_height != m_block_height + 1)
+    throw std::runtime_error("block_height should be next after the block already processed");
+
+  const StakeTransactionStorage::stake_transaction_array& stake_txs = stake_txs_storage.get_txs();
+
+    //build blockchain based list for each tier
+
+  supernode_array prev_supernodes, current_supernodes;
+  supernode_tier_array new_tier;
+
+  for (size_t i=0; i<StakeTransactionProcessor::TIERS_COUNT; i++)
+  {
+    prev_supernodes.clear();
+    current_supernodes.clear();
+
+      //prepare lists of valid supernodes (stake period is valid)
+
+    if (!m_history.empty())
+      prev_supernodes = m_history.back()[i];
+
+    std::remove_if(prev_supernodes.begin(), prev_supernodes.end(), [block_height](const supernode& desc) {
+      return !is_valid_stake(block_height, desc.block_height, desc.unlock_time);
+    });
+
+    current_supernodes.reserve(stake_txs.size());
+
+    for (const stake_transaction& stake_tx : stake_txs)
+    {
+      if (!is_valid_stake(block_height, stake_tx.block_height, stake_tx.unlock_time))
+        continue;
+
+      if (stake_tx.tier != i)
+        continue;
+
+      supernode sn;
+
+      sn.supernode_public_id = stake_tx.supernode_public_id;
+      sn.block_height        = stake_tx.block_height;
+      sn.unlock_time         = stake_tx.unlock_time;
+
+      current_supernodes.emplace_back(std::move(sn));
```

Comment:
This is minor, but this is not how `emplace_back` is meant to be used: its purpose is for in-place construction.  While you can explicitly do an in-place construction via a move construction, if you just want to move an rvalue `push_back` explicitly has an rvalue reference argument overload to handle exactly that case.

An example of how you could use `emplace_back` here to avoid a move entirely:
```C++
current_supernodes.emplace_back();
auto &sn = current_supernodes.back();
sn.supernode_public_id = stake_tx.supernode_public_id;
sn.block_height        = stake_tx.block_height;
sn.unlock_time         = stake_tx.unlock_time;
```
in C++17 this is even cleaner:
```C++
auto &sn = current_supernodes.emplace_back();
sn.supernode_public_id = stake_tx.supernode_public_id;
sn.block_height        = stake_tx.block_height;
sn.unlock_time         = stake_tx.unlock_time;
```